src/jit/codegenxarch.cpp

```diff
@@ -6252,6 +6252,51 @@ void CodeGen::genCallInstruction(GenTreePtr node)
 
     if (target != nullptr)
     {
+#ifdef _TARGET_X86_
+        if (((call->gtFlags & GTF_CALL_VIRT_KIND_MASK) == GTF_CALL_VIRT_STUB) && (call->gtCallType == CT_INDIRECT))
+        {
+            // On x86, we need to generate a very specific pattern for indirect VSD calls:
+            //
+            //    3-byte nop
+            //    call dword ptr [eax]
+            //
+            // Where EAX is also used as an argument to the stub dispatch helper. Make
+            // sure that the call target address is computed into EAX in this case.
+
+            assert(target->isContainedIndir());
+
+            // Disable random NOP emission
+            getEmitter()->emitDisableRandomNops();
+
+            GenTreeIndir* indir = target->AsIndir();
+            assert(indir->Addr() == indir->Base());
+            assert(indir->HasBase());
+            assert(!indir->HasIndex());
+            assert(indir->Scale() == 1);
+            assert(indir->Offset() == 0);
+
+            GenTree* base = indir->Base();
+            genConsumeReg(base);
```

Comment:
This brings another important question, how are we ensuring that addr in this case doesn't get marked as contained?

Two things:
1) codegen should assert here that addr is not contained
GenTree* addr = indir->Addr();
assert(!addr->isContained());

2) Looking at SetIndirAddrOpCounts() which sets opCounts on GT_IND node, there is logic to ensure that we don't mark addr as contained as follows

   else if (addr->IsCnsIntOrI() && addr->AsIntConCommon()->FitsInAddrBase(comp) &&
             addr->gtLsraInfo.getDstCandidates(m_lsra) != RBM_VIRTUAL_STUB_PARAM)


Addr on x86 is 4-bytes and hence first two conditions above will be true.
This check  requires, the LowerVirtualStubCall() to have set gtRegNum field on addr to REG_VIRTUAL_STUB_PARAM so that gtLsraInfo.Initialize() would have set dst candidates on addr.

(Note: for amd64 when we are setting dstCandiates and for x86 srcCandidates, this would mean, you would need to fix amd64 code path as well so that this check works for both code paths)


---
In reply to: [79262925](https://github.com/dotnet/coreclr/pull/7232#discussion_r79262925) [](ancestors = 79262925,79262359)